crates/matrix-sdk-common/src/ring_buffer.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{vec_deque::Iter, VecDeque};
+
+/// A simple fixed-size ring buffer implementation.
+pub struct RingBuffer<T, N: const usize> {
+    inner: VecDeque<T>,
+}
+
+impl<T, N: const usize> RingBuffer<T, N> {
+    pub fn new() -> Self {
+        Self { inner: VecDeque::with_capacity(N) }
+    }
+
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    pub fn get(&self, index: usize) -> Option<&T> {
+        self.inner.get(index)
+    }
+
+    pub fn push(&mut self, value: T) {
+        if self.inner.len() == self.inner.capacity() {
+            self.inner.pop_front();
+        }
+
+        self.inner.push_back(value);
+    }
+
+    pub fn pop(&mut self) -> Option<T> {
+        self.inner.pop_front()
+    }
+
+    pub fn iter(&self) -> Iter<'_, T> {
```

Comment:
An `IntoIterator` implementation would also be nice, but can also be added when needed.

Reply:
Yup.